src/shared/sleep-config.c

```diff
@@ -178,56 +181,69 @@ int find_hibernate_location(char **device, char **type, size_t *size, size_t *us
 
         (void) fscanf(f, "%*s %*s %*s %*s %*s\n");
 
-        // TODO: sort swaps in priority order rather than using first successful option
         for (i = 1;; i++) {
-                _cleanup_free_ char *dev_field = NULL, *type_field = NULL;
-                size_t size_field, used_field;
+                SwapEntry *swap;
                 int k;
 
+                swap = new0(SwapEntry, 1);
+                if (!swap)
+                        return log_oom();
+
                 k = fscanf(f,
                            "%ms "   /* device/file */
                            "%ms "   /* type of swap */
                            "%zu "   /* swap size */
                            "%zu "   /* used */
-                           "%*i\n", /* priority */
-                           &dev_field, &type_field, &size_field, &used_field);
+                           "%i\n",  /* priority */
+                           &swap->device, &swap->type, &swap->size, &swap->used, &swap->priority);
                 if (k == EOF)
                         break;
-                if (k != 4) {
+                if (k != 5) {
                         log_warning("Failed to parse /proc/swaps:%u", i);
                         continue;
                 }
 
-                if (streq(type_field, "file")) {
+                if (streq(swap->type, "file")) {
 
-                        if (endswith(dev_field, "\\040(deleted)")) {
-                                log_warning("Ignoring deleted swap file '%s'.", dev_field);
+                        if (endswith(swap->device, "\\040(deleted)")) {
+                                log_warning("Ignoring deleted swap file '%s'.", swap->device);
                                 continue;
                         }
 
-                } else if (streq(type_field, "partition")) {
+                } else if (streq(swap->type, "partition")) {
                         const char *fn;
 
-                        fn = path_startswith(dev_field, "/dev/");
+                        fn = path_startswith(swap->device, "/dev/");
                         if (fn && startswith(fn, "zram")) {
-                                log_debug("Ignoring compressed RAM swap device '%s'.", dev_field);
+                                log_debug("Ignoring compressed RAM swap device '%s'.", swap->device);
                                 continue;
                         }
                 }
 
-                if (device)
-                        *device = TAKE_PTR(dev_field);
-                if (type)
-                        *type = TAKE_PTR(type_field);
-                if (size)
-                        *size = size_field;
-                if (used)
-                        *used = used_field;
-                return 0;
+                if (!selected_swap || swap->priority > selected_swap->priority) {
+                        selected_swap = swap_entry_free(selected_swap);
+                        selected_swap = swap;
```

Comment:
i think it would make sense to tweak this further: for two partitions of the same priority pick the larger one. There's in fact an item in the TODO file about this, which you could then drop. 

rationale for such a logic: if the prio is the same then the larger swap device has the bigger chance of success

Reply:
This makes sense. I'll add that. It seems like perhaps it should check for partition (or file) with largest size remaining (`size - used`) rather than simply `size`?

Reply:
The other thing I'd like to add as a follow up is that I think hibernate should always prioritize what is set in `/sys/power/resume` and `/sys/power/resume_offset` over any prioritization set in `/proc/swaps`. I started to try to implement this a while back but ran into some challenges. If you think it's a reasonable idea or if you have other suggestions I'd like to take another stab at getting it working.

Reply:
yeah, checking for free space makes a lot of sense indeed

Reply:
Hmm, I thought that 2002d8cdaee105a7ab6bc89d480a9ded2198247a was supposed to deliver that /sys/power/resume when already set takes precedence?

Reply:
Thanks, @poettering 

I think the issue here is that `find_hibernate_location` has two uses - one is to identify the location where to hibernate but the other is when called by `enough_swap_for_hibernation` which uses the `used` and `size` values to compare remaining swap space to total used memory tries to determine if hibernation can succeed and this is where the potential issues reside.

I realized the other day that a user could be in the following situations if hey have multiple swaps configured and `find_hibernation_location` reports `size` and `used` on one that is not the `resume=` location (either with existing logic that just takes the first possible match in `/proc/swap` or with `priority` logic changes in progress); the `enough_swap_for_hibernation` check could report using space information for a partition/file that is not the `resume=`/`resume_offset`. In this situation there seems to be two possible bugs related to inspecting the `size` and `used` values of a swap that is not the `resume=` value: 

1. `enough_swap_to_hibernate` could be return `false` when the `resume=` swap actually does have enough space (example may be a small swap file with a high priority). This could be confusing and annoying to the user but could be worked around by setting `SYSTEMD_BYPASS_HIBERNATION_MEMORY_CHECK`
2. The other way around; `enough_swap_to_hibernate` could return `true` from when the actual `resume=` location does not have enough and hibernation will ultimately fail. In these situations, the error messages may be misleading at the point of failure. This case would fail anyway but should fail at the `enough_` check rather than at the hibernation attempt.

I think we need to make sure that when checking for `enough_swap_to_hibernate` we always check `resume=` location if it is set since this is what will be hibernated to with https://github.com/systemd/systemd/commit/2002d8cdaee105a7ab6bc89d480a9ded2198247a as you pointed out.